File: tools/preparse-dump/preparse-dump.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/// preparse-dump: Drive JSParser::preParseBuffer and print the PreParse
/// side-table in a canonical text format.  This is the C++ oracle for the
/// Rust preparse_differential test.
///
/// OUTPUT CONTRACT
///   On success: "PREPARSE <n>\n" followed by one line per entry (sorted by
///     start offset): "<start> <end> <strict> <arrow> <arrowArgs> <dirCount>
///     [dir...]\n"
///   On error: "ERROR <count>\n"
///
/// Args: <file|->
///   - means read from stdin.

#include "hermes/AST/Context.h"
#include "hermes/Parser/JSParser.h"
#include "hermes/Parser/PreParser.h"
#include "hermes/Support/SourceErrorManager.h"

#include "llvh/Support/MemoryBuffer.h"
#include "llvh/Support/raw_ostream.h"

#include <algorithm>
#include <cstdio>
#include <cstring>
#include <string>
#include <vector>

using namespace hermes;
using namespace hermes::parser;

int main(int argc, char **argv) {
  const char *filePath = nullptr;
  for (int i = 1; i < argc; ++i) {
    const char *arg = argv[i];
    if (filePath != nullptr) {
      llvh::errs() << argv[0] << ": too many arguments\n";
      return 1;
    }
    filePath = arg;
  }
  if (filePath == nullptr) {
    llvh::errs() << "Usage: " << argv[0] << " <file|->\n";
    return 1;
  }

  auto fileBufOrErr =
      llvh::MemoryBuffer::getFileOrSTDIN(llvh::StringRef(filePath));
  if (!fileBufOrErr) {
    llvh::errs() << argv[0] << ": error reading '" << filePath
                 << "': " << fileBufOrErr.getError().message() << "\n";
    return 1;
  }

  auto ctx = std::make_shared<Context>();
  SourceErrorManager &sm = ctx->getSourceErrorManager();
  // Suppress diagnostics — we report errors via "ERROR N" ourselves.
  sm.setDiagHandler([](const llvh::SMDiagnostic &, void *) {}, nullptr);

  uint32_t bufId = sm.addNewSourceBuffer(std::move(fileBufOrErr.get()));
  auto parser = JSParser::preParseBuffer(*ctx, bufId, /*strict=*/false);
  if (!parser) {
    llvh::outs() << "ERROR " << sm.getErrorCount() << "\n";
    return 0;
  }

  const char *bufStart = sm.getSourceBuffer(bufId)->getBufferStart();
  PreParsedBufferInfo *info = ctx->getPreParsedBufferInfo(bufId);

  // Collect and sort by start offset (the DenseMap is unordered).
  std::vector<std::pair<size_t, const PreParsedFunctionInfo *>> v;
  v.reserve(info->functionInfo.size());
  for (auto &kv : info->functionInfo) {
    size_t startOff = (size_t)(kv.first.getPointer() - bufStart);
    v.push_back({startOff, &kv.second});
  }
  std::sort(v.begin(), v.end(), [](auto &a, auto &b) {
    return a.first < b.first;
  });

  llvh::outs() << "PREPARSE " << v.size() << "\n";
  for (auto &e : v) {
    size_t endOff = (size_t)(e.second->end.getPointer() - bufStart);
    llvh::outs() << e.first << " " << endOff << " "
                 << (e.second->strictMode ? 1 : 0) << " "
                 << (e.second->containsArrowFunctions ? 1 : 0) << " "
                 << (e.second->mayContainArrowFunctionsUsingArguments ? 1 : 0)
                 << " " << e.second->directives.size();
    for (auto &d : e.second->directives)
      llvh::outs() << " " << d;
    llvh::outs() << "\n";
  }
  return 0;
}

File: rust/crates/tools/src/bin/preparse_dump.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! preparse-dump: Rust mirror of the C++ preparse-dump oracle tool.
//!
//! Runs a PreParse pass over a JS file and prints the side-table in a
//! canonical format. The preparse_differential test compares this byte-for-byte
//! against the C++ binary.
//!
//! OUTPUT CONTRACT
//!   On success: "PREPARSE <n>\n" then one line per entry sorted by start
//!     offset: "<start> <end> <strict> <arrow> <arrowArgs> <dirCount>
//!     [dir...]\n"
//!   On error: "ERROR <count>\n"
//!
//! Args: [--parse-flow] [--parse-ts] [file|-]
//!   omitted or "-" reads stdin.

use std::io::{self, Read, Write};

use hermes_ast::context::Context;
use hermes_parser::js::JSParserImpl;
use hermes_parser::lexer::{GrammarContext, JSLexer};
use hermes_support::manager::SourceErrorManager;

fn main() {
    let prog = "preparse-dump";
    let mut parse_flow = false;
    let mut parse_ts = false;
    let mut file_path: Option<String> = None;
    for arg in std::env::args().skip(1) {
        match arg.as_str() {
            "--parse-flow" => parse_flow = true,
            "--parse-ts" => parse_ts = true,
            _ => {
                if file_path.is_some() {
                    eprintln!("{prog}: too many arguments");
                    std::process::exit(1);
                }
                file_path = Some(arg);
            }
        }
    }
    let file_path = file_path.unwrap_or_else(|| "-".to_string());

    let bytes: Vec<u8> = if file_path == "-" {
        let mut buf = Vec::new();
        io::stdin().read_to_end(&mut buf).unwrap_or_else(|e| {
            eprintln!("{prog}: error reading stdin: {e}");
            std::process::exit(1);
        });
        buf
    } else {
        std::fs::read(&file_path).unwrap_or_else(|e| {
            eprintln!("{prog}: error reading '{file_path}': {e}");
            std::process::exit(1);
        })
    };

    let mut sm = SourceErrorManager::new();
    let buf_id = sm.add_buffer_bytes("input", &bytes);
    let mut ctx = Context::new();
    // hermesc -parse-flow defaults to ParseFlowSetting::ALL → ambiguous on
    // (same plumbing as ast_dump.rs).
    ctx.set_parse_flow(parse_flow);
    ctx.set_parse_flow_ambiguous(parse_flow);
    ctx.set_parse_ts(parse_ts);
    let gc = ctx.lock();
    let atoms = &gc.ctx().atom_table;
    let lexer = JSLexer::new(buf_id, &mut sm, atoms, GrammarContext::AllowRegExp);

    let result = JSParserImpl::pre_parse_buffer(&gc, lexer, false);

    let stdout = io::stdout();
    let mut out = stdout.lock();

    match result {
        None => {
            writeln!(out, "ERROR {}", sm.error_count()).unwrap();
        }
        Some(mut parser) => {
            let info = parser.take_pre_parsed();
            // Sort entries by start offset for a canonical ordering.
            let mut entries: Vec<(u32, &_)> =
                info.function_info.iter().map(|(k, v)| (*k, v)).collect();
            entries.sort_by_key(|(k, _)| *k);

            writeln!(out, "PREPARSE {}", entries.len()).unwrap();
            for (start, fi) in &entries {
                write!(
                    out,
                    "{} {} {} {} {} {}",
                    start,
                    fi.end.offset,
                    fi.strict_mode as u8,
                    fi.contains_arrow_functions as u8,
                    fi.may_contain_arrow_functions_using_arguments as u8,
                    fi.directives.len(),
                )
                .unwrap();
                for d in &fi.directives {
                    write!(out, " ").unwrap();
                    out.write_all(d).unwrap();
                }
                writeln!(out).unwrap();
            }
        }
    }
    out.flush().ok();
}
